openSearchLane.go

```diff
@@ -0,0 +1,425 @@
+	emergencyFn    EmergencyFn
+	ctx            context.Context
+	cancelFn       context.CancelFunc
+	messagesQueued int
+	messagesSent   int
+}
+
+type OpenSearchLane interface {
+	Lane
+	LaneMetadata
+	Connect(config OpenSearchLaneConfig) (err error)
+	SetEmergencyHandler(emergencyFn EmergencyFn)
+	Stats()
+}
+
+func NewOpenSearchLane(ctx context.Context) (l OpenSearchLane) {
+	ll := deriveLogLane(nil, ctx, []Lane{}, "")
+
+	ctx, cancelFn := context.WithCancel(context.Background())
+
+	osl := openSearchLane{
+		openSearchConnection: &openSearchConnection{
+			logBuffer: make([]*OpenSearchLogMessage, 0),
+			stopCh:    make(chan *sync.WaitGroup, 1),
+			wakeCh:    make(chan struct{}, 1),
+			ctx:       ctx,
+			cancelFn:  cancelFn,
+		},
+	}
+	osl.openSearchConnection.attach()
+
+	ll.clone(&osl.logLane)
+	osl.logLane.writer = log.New(&osl, "", 0)
+	osl.wlog.SetFlags(log.Flags() &^ (log.Ldate | log.Ltime))
+
+	l = &osl
+
+	return
+
+}
+
+func (osl *openSearchLane) Connect(config OpenSearchLaneConfig) (err error) {
+	if osl.openSearchConnection.client != nil {
+		osl.Warn("there is already a connection to OpenSearch")
+		return
+	}
+
+	client, err := newOpenSearchClient(config.OpenSearchUrl, config.OpenSearchPort, config.OpenSearchUser, config.OpenSearchPass)
+	if err != nil {
+		return
+	}
+	osl.openSearchConnection.client = client
+	osl.openSearchConnection.config = config
+
+	if config.LogThreshold == 0 {
+		osl.openSearchConnection.config.LogThreshold = DefaultLogThreshold
+	}
+	if config.MaxBufferSize == 0 {
+		osl.openSearchConnection.config.MaxBufferSize = DefaultMaxBufferSize
+	}
+	if config.BackoffInterval == 0 {
+		osl.openSearchConnection.config.BackoffInterval = DefaultBackoffInterval
+	}
+	if config.BackOffLimit == 0 {
+		osl.openSearchConnection.config.BackOffLimit = DefaultBackOffLimit
+	}
+
+	go osl.openSearchConnection.processConnection(osl)
+
+	return
+
+}
+
+func (osl *openSearchLane) Derive() Lane {
+	ll := deriveLogLane(&osl.logLane, context.WithValue(osl.Context, parent_lane_id, osl.LaneId()), osl.tees, osl.cr)
+
+	osl2 := openSearchLane{
+		openSearchConnection: osl.openSearchConnection,
+	}
+	osl.openSearchConnection.attach()
+
+	ll.clone(&osl2.logLane)
+	osl2.logLane.writer = log.New(&osl2, "", 0)
+
+	return &osl2
+}
+
+func (osl *openSearchLane) DeriveWithCancel() (Lane, context.CancelFunc) {
+	childCtx, cancelFn := context.WithCancel(context.WithValue(osl.logLane.Context, parent_lane_id, osl.logLane.LaneId()))
+	ll := deriveLogLane(&osl.logLane, childCtx, osl.tees, osl.cr)
+
+	osl2 := openSearchLane{
+		openSearchConnection: osl.openSearchConnection,
+	}
+	osl.openSearchConnection.attach()
+
+	ll.clone(&osl2.logLane)
+	osl2.logLane.writer = log.New(&osl2, "", 0)
+
+	return &osl2, cancelFn
+}
+
+func (osl *openSearchLane) DeriveReplaceContext(ctx context.Context) Lane {
+	ll := deriveLogLane(&osl.logLane, context.WithValue(ctx, parent_lane_id, osl.LaneId()), osl.tees, osl.cr)
+
+	osl2 := openSearchLane{
```

Comment:
minor, it seems these 3 are doing the exact same thing after the deriveLogLane call and could be DRY with something like `finishDerive()`